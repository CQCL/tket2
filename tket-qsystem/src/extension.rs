//! This module defines the Hugr extensions used by tket-qsystem.
pub mod futures;
pub mod qsystem;
pub mod random;
pub mod result;
pub mod utils;
pub mod wasm;
