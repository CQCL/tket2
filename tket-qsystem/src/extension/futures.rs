//! This module defines the Hugr extension used to represent futures.
//!
//! `Future<t>` is a linear type representing a value that will be available in
//! the future.  It can be consumed by `Read`, returning a `t`.  It can be
//! duplicated by `Dup`, and discarded with `Free`.
use std::sync::{Arc, Weak};

use hugr::{
    builder::{BuildError, Dataflow},
    extension::{
        simple_op::{
            try_from_name, HasConcrete, HasDef, MakeExtensionOp, MakeOpDef, MakeRegisteredOp,
            OpLoadError,
        },
        ExtensionBuildError, ExtensionId, OpDef, SignatureError, SignatureFunc, TypeDef, Version,
    },
    ops::{custom::ExtensionOp, OpType},
    types::{type_param::TypeParam, CustomType, PolyFuncType, Signature, Type, TypeArg, TypeBound},
    Extension, Wire,
};
use lazy_static::lazy_static;
use serde::{Deserialize, Serialize};
use smol_str::SmolStr;
use strum::{EnumIter, EnumString, IntoStaticStr};

/// The ID of the `tket.futures` extension.
pub const EXTENSION_ID: ExtensionId = ExtensionId::new_unchecked("tket.futures");
/// The "tket.futures" extension version
pub const EXTENSION_VERSION: Version = Version::new(0, 2, 0);

lazy_static! {
    /// The "tket.futures" extension.
    pub static ref EXTENSION: Arc<Extension>  = {
        Extension::new_arc(EXTENSION_ID, EXTENSION_VERSION, |ext, ext_ref| {
            let _ = add_future_type_def(ext, ext_ref.clone()).unwrap();

            FutureOpDef::load_all_ops( ext, ext_ref).unwrap();
        })
    };

    /// The name of the `Future` type.
    pub static ref FUTURE_TYPE_NAME: SmolStr = SmolStr::new_inline("Future");
}

fn add_future_type_def(
    ext: &mut Extension,
    extension_ref: Weak<Extension>,
) -> Result<&TypeDef, ExtensionBuildError> {
    ext.add_type(
        FUTURE_TYPE_NAME.to_owned(),
        vec![TypeBound::Linear.into()],
        "A value that is computed asynchronously".into(),
        TypeBound::Linear.into(),
        &extension_ref,
    )
}

/// Returns a `Future<t>` [CustomType].
pub fn future_custom_type(t: Type, extension_ref: &Weak<Extension>) -> CustomType {
    CustomType::new(
        FUTURE_TYPE_NAME.to_owned(),
        vec![t.into()],
        EXTENSION_ID,
        TypeBound::Linear,
        extension_ref,
    )
}

/// Returns a `Future<t>` [Type].
pub fn future_type(t: Type) -> Type {
    future_custom_type(t, &Arc::downgrade(&EXTENSION)).into()
}

#[derive(
    Clone,
    Copy,
    Debug,
    Serialize,
    Deserialize,
    Hash,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    EnumIter,
    IntoStaticStr,
    EnumString,
)]
#[allow(missing_docs)]
#[non_exhaustive]
/// Simple enum of "tket.futures" operations.
pub enum FutureOpDef {
    Read,
    Dup,
    Free,
}

impl MakeOpDef for FutureOpDef {
    fn opdef_id(&self) -> hugr::ops::OpName {
        <&'static str>::from(self).into()
    }

    fn init_signature(&self, extension_ref: &Weak<Extension>) -> SignatureFunc {
        let t_param = TypeParam::from(TypeBound::Linear);
        let t_type = Type::new_var_use(0, TypeBound::Linear);
        let future_type = Type::new_extension(future_custom_type(t_type.clone(), extension_ref));
        match self {
            FutureOpDef::Read => {
                PolyFuncType::new([t_param], Signature::new(future_type, t_type)).into()
            }
            FutureOpDef::Dup => PolyFuncType::new(
                [t_param],
                Signature::new(future_type.clone(), vec![future_type.clone(), future_type]),
            )
            .into(),
            FutureOpDef::Free => {
                PolyFuncType::new([t_param], Signature::new(future_type.clone(), vec![])).into()
            }
        }
    }

    fn extension(&self) -> ExtensionId {
        EXTENSION_ID
    }

    fn from_def(op_def: &OpDef) -> Result<Self, hugr::extension::simple_op::OpLoadError> {
        try_from_name(op_def.name(), op_def.extension_id())
    }

    fn description(&self) -> String {
        match self {
            FutureOpDef::Read => "Read a value from a Future, consuming it".into(),
            FutureOpDef::Dup => {
                "Duplicate a Future. The original Future is consumed and two Futures are returned"
                    .into()
            }
            FutureOpDef::Free => "Consume a future without reading it.".into(),
        }
    }

    fn extension_ref(&self) -> Weak<Extension> {
        Arc::downgrade(&EXTENSION)
    }
}

impl HasConcrete for FutureOpDef {
    type Concrete = FutureOp;

    fn instantiate(&self, type_args: &[TypeArg]) -> Result<Self::Concrete, OpLoadError> {
        match type_args {
            [TypeArg::Runtime(ty)] => Ok(FutureOp {
                op: *self,
                typ: ty.clone(),
            }),
            _ => Err(SignatureError::InvalidTypeArgs.into()),
        }
    }
}

impl From<&FutureOp> for &'static str {
    fn from(value: &FutureOp) -> Self {
        value.op.into()
    }
}

/// Concrete "tket.futures" operations with type set.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct FutureOp {
    /// The `FutureOpDef` that defines this operation.
    pub op: FutureOpDef,
    /// The inner type of the `Future` this op acts on.
    pub typ: Type,
}

impl MakeExtensionOp for FutureOp {
    fn op_id(&self) -> hugr::ops::OpName {
        self.op.opdef_id()
    }

    fn from_extension_op(ext_op: &ExtensionOp) -> Result<Self, OpLoadError>
    where
        Self: Sized,
    {
        FutureOpDef::from_def(ext_op.def())?.instantiate(ext_op.args())
    }

    fn type_args(&self) -> Vec<hugr::types::TypeArg> {
        vec![self.typ.clone().into()]
    }
}

impl MakeRegisteredOp for FutureOp {
    fn extension_id(&self) -> ExtensionId {
        EXTENSION_ID
    }

    fn extension_ref(&self) -> Weak<Extension> {
        Arc::downgrade(&EXTENSION)
    }
}

impl HasDef for FutureOp {
    type Def = FutureOpDef;
}

impl TryFrom<&OpType> for FutureOpDef {
    type Error = OpLoadError;

    fn try_from(value: &OpType) -> Result<Self, Self::Error> {
        Self::from_op(
            value
                .as_extension_op()
                .ok_or(OpLoadError::NotMember(value.to_string()))?,
        )
    }
}

/// An extension trait for [Dataflow] providing methods to add "tket.futures"
/// operations.
pub trait FutureOpBuilder: Dataflow {
    /// Add a "tket.futures.Read" op.
    fn add_read(&mut self, lifted: Wire, typ: Type) -> Result<[Wire; 1], BuildError> {
        Ok(self
            .add_dataflow_op(
                FutureOp {
                    op: FutureOpDef::Read,
                    typ,
                },
                [lifted],
            )?
            .outputs_arr())
    }

    /// Add a "tket.futures.Dup" op.
    fn add_dup(&mut self, lifted: Wire, typ: Type) -> Result<[Wire; 2], BuildError> {
        Ok(self
            .add_dataflow_op(
                FutureOp {
                    op: FutureOpDef::Dup,
                    typ,
                },
                [lifted],
            )?
            .outputs_arr())
    }

    /// Add a "tket.futures.Free" op.
    fn add_free(&mut self, lifted: Wire, typ: Type) -> Result<(), BuildError> {
        let op = self.add_dataflow_op(
            FutureOp {
                op: FutureOpDef::Free,
                typ,
            },
            [lifted],
        )?;
        assert!(op.outputs().len() == 0);
        Ok(())
    }
}

impl<D: Dataflow> FutureOpBuilder for D {}

#[cfg(test)]
pub(crate) mod test {

    use hugr::builder::{Dataflow, DataflowHugr, FunctionBuilder};
    use hugr::HugrView;
    use std::sync::Arc;
    use strum::IntoEnumIterator;

    use super::*;

    fn get_opdef(op: FutureOpDef) -> Option<&'static Arc<OpDef>> {
        EXTENSION.get_op(&op.opdef_id())
    }

    #[test]
    fn create_extension() {
        assert_eq!(EXTENSION.name(), &EXTENSION_ID);

        for o in FutureOpDef::iter() {
            assert_eq!(FutureOpDef::from_def(get_opdef(o).unwrap()), Ok(o));
        }
    }

    #[test]
    fn future_op_from_def() {
        let typ = Type::UNIT;

        assert_eq!(
            FutureOp {
                op: FutureOpDef::Free,
                typ: typ.clone()
            },
            FutureOpDef::Free.instantiate(&[typ.into()]).unwrap()
        )
    }

    #[test]
    fn circuit() {
        let t_param = TypeParam::from(TypeBound::Linear);
        let t = Type::new_var_use(0, TypeBound::Linear);
        let future_type = future_type(t.clone());

        let hugr = {
            let mut func_builder = FunctionBuilder::new(
                "circuit",
                PolyFuncType::new(vec![t_param], Signature::new(future_type, t.clone())),
            )
            .unwrap();
            let [future_w] = func_builder.input_wires_arr();
            let [future_w, lazy_dup_w] = func_builder.add_dup(future_w, t.clone()).unwrap();
            func_builder.add_free(future_w, t.clone()).unwrap();
            let [t_w] = func_builder.add_read(lazy_dup_w, t).unwrap();
            func_builder.finish_hugr_with_outputs([t_w]).unwrap()
        };
        hugr.validate().unwrap();
    }
}
