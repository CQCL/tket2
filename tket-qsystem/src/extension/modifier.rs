use hugr::{Hugr, HugrView};

use crate::QSystemPassError;
use tket::rich_circuit::modifier_resolver::resolve_modifier_with_entrypoints;

#[allow(missing_docs)]
#[derive(Default)]
pub struct ModifierResolverPass;

impl ModifierResolverPass {
    /// Run ModifierResolverPass.
    pub fn run(self, hugr: &mut Hugr) -> Result<(), QSystemPassError> {
        resolve_modifier_with_entrypoints(hugr, [hugr.entrypoint()]).unwrap();
        return Ok(());
    }
}
