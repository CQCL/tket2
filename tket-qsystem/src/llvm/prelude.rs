//! LLVM lowering implementations for HUGR prelude operations.

use std::hash::{DefaultHasher, Hash as _, Hasher as _};
use tket::hugr::{self, llvm::inkwell};

use anyhow::{anyhow, ensure, Result};
use hugr::llvm::emit::func::EmitFuncContext;
use hugr::llvm::extension::prelude::PreludeCodegen;
use hugr::llvm::types::TypingSession;
use inkwell::attributes::{Attribute, AttributeLoc};
use inkwell::module::Linkage;
use inkwell::types::{BasicType, IntType};
use inkwell::values::{BasicValue, BasicValueEnum, FunctionValue, StructValue};
use inkwell::AddressSpace;
use tket::hugr::{HugrView, Node};

#[derive(Clone)]
/// Codegen extension for the QIS prelude.
pub struct QISPreludeCodegen;

impl PreludeCodegen for QISPreludeCodegen {
    fn usize_type<'c>(&self, session: &TypingSession<'c, '_>) -> IntType<'c> {
        session.iw_context().i64_type()
    }

    /// Return the llvm type of [`tket::hugr::extension::prelude::qb_t`].
    fn qubit_type<'c>(&self, session: &TypingSession<'c, '_>) -> impl BasicType<'c> {
        session.iw_context().i64_type()
    }

    /// A panic ends all shots, this is done by using a return code > 1000.
    /// The lowering of the panic HUGR op shifts the return code held by the op by 1000.
    /// So valid values for `signal` in the panic HUGR op are 1-(2^31 - 1001).
    fn emit_panic<H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<H>,
        err: BasicValueEnum,
    ) -> Result<()> {
        let err = StructValue::try_from(err)
            .map_err(|()| anyhow!("emit_panic: Expected err to have struct type"))?;
        let return_code = ctx.builder().build_extract_value(err, 0, "")?;
        // TODO check if code is in valid range 1-(2^31 - 1001), otherwise error.
        // https://github.com/quantinuum-dev/eldarion/issues/287
        let return_code = {
            let shift = ctx.iw_context().i32_type().const_int(1000, false);
            ctx.builder()
                .build_int_add(return_code.into_int_value(), shift, "shift_code")?
        };

        let msg = ctx.builder().build_extract_value(err, 1, "")?;
        let panic = Self::get_panic(ctx)?;
        ctx.builder()
            .build_call(panic, &[return_code.into(), msg.into()], "")?;
        Ok(())
    }

    /// An exit ends the current shot using a return code `rc`, where `1 <= rc <= 1000`.
    fn emit_exit<H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<H>,
        err: BasicValueEnum,
    ) -> Result<()> {
        let err = StructValue::try_from(err)
            .map_err(|()| anyhow!("emit_exit: Expected err to have struct type"))?;
        let return_code = ctx.builder().build_extract_value(err, 0, "")?;
        // TODO check if code is in valid range 1-1000, otherwise error.
        // https://github.com/quantinuum-dev/eldarion/issues/287
        let msg = ctx.builder().build_extract_value(err, 1, "")?;
        let panic = Self::get_panic(ctx)?;
        ctx.builder()
            .build_call(panic, &[return_code.into(), msg.into()], "")?;
        Ok(())
    }

    fn emit_print<H: HugrView<Node = Node>>(
        &self,
        _ctx: &mut EmitFuncContext<H>,
        _text: inkwell::values::BasicValueEnum,
    ) -> Result<()> {
        Ok(())
    }

    fn emit_const_error<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        err: &hugr::extension::prelude::ConstError,
    ) -> Result<BasicValueEnum<'c>> {
        let err_ty = ctx
            .llvm_type(&hugr::extension::prelude::error_type())?
            .into_struct_type();
        let signal = err_ty
            .get_field_type_at_index(0)
            .unwrap()
            .into_int_type()
            .const_int(u64::from(err.signal), false);

        let message = emit_global_string(ctx, &err.message, "e_", "EXIT:INT:")?;
        let err = err_ty.const_named_struct(&[signal.into(), message]);
        Ok(err.into())
    }

    fn emit_const_string<'c, H: HugrView<Node = Node>>(
        &self,
        ctx: &mut EmitFuncContext<'c, '_, H>,
        str: &hugr::extension::prelude::ConstString,
    ) -> Result<BasicValueEnum<'c>> {
        emit_global_string(ctx, str.value(), "s_", "")
    }
}

impl QISPreludeCodegen {
    fn get_panic<'c, H: HugrView<Node = Node>>(
        ctx: &EmitFuncContext<'c, '_, H>,
    ) -> Result<FunctionValue<'c>> {
        let iwc = ctx.iw_context();
        let func_type = iwc.void_type().fn_type(
            &[
                iwc.i32_type().into(),
                iwc.i8_type().ptr_type(AddressSpace::default()).into(),
            ],
            false,
        );
        let f = ctx.get_extern_func("panic", func_type)?;
        let noreturn = Attribute::get_named_enum_kind_id("noreturn");
        debug_assert!(noreturn != 0);
        f.add_attribute(
            AttributeLoc::Function,
            iwc.create_enum_attribute(noreturn, 0),
        );

        Ok(f)
    }
}

/// Emit a global string constant with a unique name.
///
/// The symbol name is constructed with `symbol_prefix` and ``str`.
///
/// The string is prefixed with `str_prefix`.
pub fn emit_global_string<'c, H: HugrView<Node = Node>>(
    ctx: &EmitFuncContext<'c, '_, H>,
    str: impl AsRef<str>,
    symbol_prefix: impl AsRef<str>,
    str_prefix: impl AsRef<str>,
) -> Result<BasicValueEnum<'c>> {
    let (str, symbol_prefix, str_prefix) =
        (str.as_ref(), symbol_prefix.as_ref(), str_prefix.as_ref());
    let tagged_str = format!("{str_prefix}{str}");
    let tagged_str_bytes = tagged_str.as_bytes();
    let tagged_str_len = tagged_str_bytes.len();
    ensure!(
        tagged_str_len < 256,
        "Constant string too long: {tagged_str_len} >= 256"
    );

    let tgt_str_bytes = [&[u8::try_from(tagged_str_len)?], tagged_str_bytes].concat();
    let tgt_str_value = ctx.iw_context().const_string(&tgt_str_bytes, false);

    let global = {
        // we need a unique name for the global. We use a prefix(of length
        // 10) of the string, followed by the hash of the whole string.
        //
        // This should be enough, but for absolute guarantee of success we
        // append ".i", i an integer and find an available name.
        let hash = {
            let mut s = DefaultHasher::default();
            tgt_str_bytes.hash(&mut s);
            let h = s.finish();
            let lower32 = h as u32;
            format!("{lower32:08X}")
        };
        // Use unicode-aware character slicing to prevent slicing through multi-byte characters
        let prefix_chars: String = str.chars().take(10).collect();
        let name_prefix = format!("{symbol_prefix}{prefix_chars}.{hash}");

        (0..)
            .find_map(|i| {
                let name = format!("{name_prefix}.{i}");
                if let Some(global) = ctx.get_current_module().get_global(&name) {
                    if let Some(v) = global.get_initializer() {
                        if v == tgt_str_value {
                            // We've found a global with the same value, use that.
                            return Some(global);
                        }
                    }
                } else {
                    let global =
                        ctx.get_current_module()
                            .add_global(tgt_str_value.get_type(), None, &name);
                    global.set_initializer(&tgt_str_value);
                    global.set_linkage(Linkage::Private);
                    global.set_constant(true);
                    return Some(global);
                }
                None
            })
            .unwrap()
    };
    Ok(ctx
        .builder()
        .build_pointer_cast(
            global.as_pointer_value(),
            ctx.iw_context().i8_type().ptr_type(AddressSpace::default()),
            "",
        )?
        .as_basic_value_enum())
}

#[cfg(test)]
mod test {
    use hugr::builder::{Dataflow, DataflowHugr};
    use hugr::extension::prelude::{self, qb_t, ConstError, EXIT_OP_ID, PANIC_OP_ID};
    use hugr::extension::PRELUDE;
    use hugr::llvm::check_emission;
    use hugr::llvm::emit::test::SimpleHugrConfig;

    use hugr::llvm::test::llvm_ctx;
    use hugr::types::TypeArg;
    use rstest::rstest;

    use super::*;

    #[rstest]
    fn test_panic_emit() {
        let mut llvm_ctx = llvm_ctx(0);
        let prelude_codegen = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| ceb.add_prelude_extensions(prelude_codegen.clone()));

        // Create a hugr that has a panic message
        let error_val = ConstError::new(42, "PANIC");
        let type_arg_q = TypeArg::Runtime(qb_t());
        let type_arg_2q = TypeArg::List(vec![type_arg_q.clone(), type_arg_q]);
        let panic_op = PRELUDE
            .instantiate_extension_op(&PANIC_OP_ID, [type_arg_2q.clone(), type_arg_2q.clone()])
            .unwrap();

        let hugr = SimpleHugrConfig::new()
            .with_ins(vec![qb_t(), qb_t()])
            .with_outs(vec![qb_t(), qb_t()])
            .with_extensions(prelude::PRELUDE_REGISTRY.to_owned())
            .finish(|mut builder| {
                let [q0, q1] = builder.input_wires_arr();
                let err = builder.add_load_value(error_val);
                let [q0, q1] = builder
                    .add_dataflow_op(panic_op, [err, q0, q1])
                    .unwrap()
                    .outputs_arr();
                builder.finish_hugr_with_outputs([q0, q1]).unwrap()
            });

        check_emission!(hugr, llvm_ctx);
    }

    #[rstest]
    fn test_exit_emit() {
        let mut llvm_ctx = llvm_ctx(0);
        let prelude_codegen = QISPreludeCodegen;
        llvm_ctx.add_extensions(move |ceb| ceb.add_prelude_extensions(prelude_codegen.clone()));

        let error_val = ConstError::new(42, "EXIT");
        let type_arg_q: TypeArg = TypeArg::Runtime(qb_t());
        let type_arg_2q: TypeArg = TypeArg::List(vec![type_arg_q.clone(), type_arg_q]);
        let exit_op = PRELUDE
            .instantiate_extension_op(&EXIT_OP_ID, [type_arg_2q.clone(), type_arg_2q.clone()])
            .unwrap();

        let hugr = SimpleHugrConfig::new()
            .with_ins(vec![qb_t(), qb_t()])
            .with_outs(vec![qb_t(), qb_t()])
            .with_extensions(prelude::PRELUDE_REGISTRY.to_owned())
            .finish(|mut builder| {
                let [q0, q1] = builder.input_wires_arr();
                let err = builder.add_load_value(error_val);
                let [q0, q1] = builder
                    .add_dataflow_op(exit_op, [err, q0, q1])
                    .unwrap()
                    .outputs_arr();
                builder.finish_hugr_with_outputs([q0, q1]).unwrap()
            });

        check_emission!(hugr, llvm_ctx);
    }
}
