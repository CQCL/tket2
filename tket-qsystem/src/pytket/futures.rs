//! Encoder and decoder for floating point operations.

use hugr::extension::simple_op::MakeExtensionOp;
use hugr::extension::ExtensionId;
use hugr::ops::ExtensionOp;
use hugr::types::Term;
use hugr::HugrView;
use itertools::Itertools;
use tket::serialize::pytket::encoder::EncodeStatus;
use tket::serialize::pytket::extension::{PytketTypeTranslator, RegisterCount};
use tket::serialize::pytket::{
    PytketEmitter, PytketEncodeError, PytketEncoderContext, TypeTranslatorSet,
};
use tket::Circuit;

use crate::extension::futures::{self, FutureOpDef};

/// Emitter for [futures](crate::extension::futures) operations and types.
///
/// The `Future<T>` type is treated as a transparent wrapper when translated to
/// pytket circuits, and operations dealing with futures do not produce pytket
/// commands.
#[derive(Debug, Clone, Default)]
pub struct FutureEmitter;

impl<H: HugrView> PytketEmitter<H> for FutureEmitter {
    fn extensions(&self) -> Option<Vec<ExtensionId>> {
        Some(vec![futures::EXTENSION_ID])
    }

    fn op_to_pytket(
        &self,
        node: H::Node,
        op: &ExtensionOp,
        circ: &Circuit<H>,
        encoder: &mut PytketEncoderContext<H>,
    ) -> Result<EncodeStatus, PytketEncodeError<H::Node>> {
        let Ok(rot_op) = FutureOpDef::from_extension_op(op) else {
            return Ok(EncodeStatus::Unsupported);
        };

        match rot_op {
            FutureOpDef::Read => {
                // Transparent map
                encoder.emit_transparent_node(node, circ, |ps| ps.input_params.to_vec())?;
                Ok(EncodeStatus::Success)
            }
            FutureOpDef::Dup => {
                // Register the same input values for each output.
                let values = encoder.get_input_values(node, circ)?;
                let outputs = circ.hugr().node_outputs(node).collect_vec();
                let out0 = hugr::Wire::new(node, outputs[0]);
                let out1 = hugr::Wire::new(node, outputs[1]);

                encoder.values.register_wire(out0, values.clone(), circ)?;
                encoder.values.register_wire(out1, values, circ)?;

                Ok(EncodeStatus::Success)
            }
            FutureOpDef::Free => Ok(EncodeStatus::Success),
        }
    }
}

impl PytketTypeTranslator for FutureEmitter {
    fn extensions(&self) -> Vec<ExtensionId> {
        vec![futures::EXTENSION_ID]
    }

    fn type_to_pytket(
        &self,
        typ: &hugr::types::CustomType,
        type_translators: &TypeTranslatorSet,
    ) -> Option<RegisterCount> {
        if typ.name() != futures::FUTURE_TYPE_NAME.as_str() {
            return None;
        }
        let Some(Term::Runtime(inner_ty)) = typ.args().first() else {
            return None;
        };

        type_translators.type_to_pytket(inner_ty)
    }
}
