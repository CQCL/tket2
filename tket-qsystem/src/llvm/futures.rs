//! LLVM lowering implementations for "tket.futures" operations.

use crate::extension::futures::{self, FutureOp, FutureOpDef, FUTURE_TYPE_NAME};
use anyhow::{anyhow, Result};
use hugr::extension::prelude::bool_t;
use hugr::extension::simple_op::MakeExtensionOp;
use hugr::ops::{ExtensionOp, Value};
use hugr::std_extensions::arithmetic::int_types::INT_TYPES;
use hugr::types::TypeArg;
use hugr::{HugrView, Node};
use hugr_llvm::custom::CodegenExtension;
use hugr_llvm::emit::func::EmitFuncContext;
use hugr_llvm::emit::{emit_value, EmitOpArgs};
use hugr_llvm::CodegenExtsBuilder;
use inkwell::builder::Builder;
use inkwell::context::Context;
use inkwell::types::{BasicTypeEnum, IntType};
use inkwell::values::FunctionValue;
use tket::hugr::{
    self,
    llvm::{self as hugr_llvm, inkwell},
};

/// Codegen extension for futures
pub struct FuturesCodegenExtension;

impl CodegenExtension for FuturesCodegenExtension {
    fn add_extension<'a, H: HugrView<Node = Node> + 'a>(
        self,
        builder: CodegenExtsBuilder<'a, H>,
    ) -> CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder
            .custom_type(
                (futures::EXTENSION_ID, FUTURE_TYPE_NAME.to_owned()),
                |session, hugr_type| {
                    match (hugr_type.name().as_str(), hugr_type.args()) {
                        // For now, we only support future bools
                        ("Future", [TypeArg::Runtime(ty)]) if *ty == bool_t() => {
                            Ok(future_type(session.iw_context()))
                        }
                        ("Future", [TypeArg::Runtime(ty)]) if *ty == INT_TYPES[6] => {
                            Ok(future_type(session.iw_context()))
                        }
                        _ => Err(anyhow!(
                            "FuturesCodegenExtension: Unsupported type: {}",
                            hugr_type
                        )),
                    }
                },
            )
            .simple_extension_op::<FutureOpDef>(|context, args, _op| {
                FuturesEmitter(context).emit(args)
            })
    }
}

/// The LLVM type representing a future handle.
pub fn future_type(context: &Context) -> BasicTypeEnum<'_> {
    // The runtime represent a future handle as an i64
    context.i64_type().into()
}

struct FuturesEmitter<'c, 'd, 'e, H: HugrView<Node = Node>>(&'d mut EmitFuncContext<'c, 'e, H>);

impl<'c, H: HugrView<Node = Node>> FuturesEmitter<'c, '_, '_, H> {
    fn iw_context(&self) -> &'c Context {
        self.0.typing_session().iw_context()
    }

    fn ll_future_type(&self) -> BasicTypeEnum<'c> {
        future_type(self.iw_context())
    }

    fn ll_bool_type(&self) -> IntType<'c> {
        self.iw_context().bool_type()
    }

    fn ll_uint_type(&self) -> IntType<'c> {
        self.iw_context().i64_type()
    }

    fn builder(&self) -> &Builder<'c> {
        self.0.builder()
    }

    fn get_func_inc_refcount(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .iw_context()
            .void_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___inc_future_refcount", func_type)
    }

    fn get_func_dec_refcount(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .iw_context()
            .void_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___dec_future_refcount", func_type)
    }

    fn get_func_read_bool(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .ll_bool_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___read_future_bool", func_type)
    }

    fn get_func_read_uint(&self) -> Result<FunctionValue<'c>> {
        let func_type = self
            .ll_uint_type()
            .fn_type(&[self.ll_future_type().into()], false);
        self.0.get_extern_func("___read_future_uint", func_type)
    }

    fn emit(&mut self, args: EmitOpArgs<'c, '_, ExtensionOp, H>) -> Result<()> {
        let future_op = FutureOp::from_optype(&args.node().generalise()).unwrap();
        let op = &future_op.op;
        let typ = &future_op.typ;
        match (op, typ) {
            (FutureOpDef::Dup, _) => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Dup expects a single input"))?;
                let func = self.get_func_inc_refcount()?;
                self.builder()
                    .build_call(func, &[arg.into()], "inc_refcount")?;
                args.outputs.finish(self.builder(), [arg, arg])
            }
            (FutureOpDef::Free, _) => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Free expects a single input"))?;
                let func = self.get_func_dec_refcount()?;
                self.builder()
                    .build_call(func, &[arg.into()], "dec_refcount")?;
                args.outputs.finish(self.builder(), [])
            }
            (FutureOpDef::Read, ty) if *ty == bool_t() => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Read expects a single input"))?;
                let read_func = self.get_func_read_bool()?;
                let result_i1 = self
                    .builder()
                    .build_call(read_func, &[arg.into()], "read_bool")?
                    .try_as_basic_value()
                    .unwrap_left()
                    .into_int_value();
                let dec_refcount_func = self.get_func_dec_refcount()?;
                self.builder()
                    .build_call(dec_refcount_func, &[arg.into()], "dec_refcount")?;
                let true_val = emit_value(self.0, &Value::true_val())?;
                let false_val = emit_value(self.0, &Value::false_val())?;
                let result = self
                    .builder()
                    .build_select(result_i1, true_val, false_val, "measure")?;
                args.outputs.finish(self.builder(), [result])
            }
            (FutureOpDef::Read, ty) if *ty == INT_TYPES[6] => {
                let [arg] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("Read expects a single input"))?;
                let read_func = self.get_func_read_uint()?;
                let result = self
                    .builder()
                    .build_call(read_func, &[arg.into()], "read_uint")?
                    .try_as_basic_value()
                    .unwrap_left();
                let dec_refcount_func = self.get_func_dec_refcount()?;
                self.builder()
                    .build_call(dec_refcount_func, &[arg.into()], "dec_refcount")?;
                args.outputs.finish(self.builder(), [result])
            }
            _ => Err(anyhow!(
                "Unsupported future operation: {:?} with type: {}",
                op,
                typ
            )),
        }
    }
}

#[cfg(test)]
mod test {
    use hugr::extension::simple_op::MakeRegisteredOp;
    use hugr::std_extensions::arithmetic::int_types::int_type;
    use hugr_llvm::check_emission;
    use hugr_llvm::test::llvm_ctx;
    use hugr_llvm::test::single_op_hugr;
    use hugr_llvm::test::TestContext;

    use super::*;
    #[rstest::rstest]
    #[case::read_bool(1,FutureOp { op: FutureOpDef::Read, typ: bool_t() })]
    #[case::dup_bool(2,FutureOp { op: FutureOpDef::Dup, typ: bool_t() })]
    #[case::free_bool(3,FutureOp { op: FutureOpDef::Free, typ: bool_t() })]
    #[case::read_int(4,FutureOp { op: FutureOpDef::Read, typ: int_type(6) })]
    #[case::dup_int(5,FutureOp { op: FutureOpDef::Dup, typ: int_type(6) })]
    #[case::free_int(6,FutureOp { op: FutureOpDef::Free, typ: int_type(6) })]
    fn emit_futures_codegen(
        #[case] _i: i32,
        #[with(_i)] mut llvm_ctx: TestContext,
        #[case] op: FutureOp,
    ) {
        llvm_ctx.add_extensions(|ceb| {
            ceb.add_extension(FuturesCodegenExtension)
                .add_default_int_extensions()
        });
        let hugr = single_op_hugr(op.to_extension_op().unwrap().into());
        check_emission!(hugr, llvm_ctx);
    }
}
