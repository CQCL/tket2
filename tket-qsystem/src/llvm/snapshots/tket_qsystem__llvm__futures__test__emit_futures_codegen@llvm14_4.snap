---
source: tket-qsystem/src/llvm/futures.rs
expression: mod_str
---
; ModuleID = 'test_context'
source_filename = "test_context"

define private i64 @_hl.main.1(i64 %0) {
alloca_block:
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  %read_uint = call i64 @___read_future_uint(i64 %0)
  call void @___dec_future_refcount(i64 %0)
  ret i64 %read_uint
}

declare i64 @___read_future_uint(i64)

declare void @___dec_future_refcount(i64)
